### PR TITLE
Conv3d

### PR DESCRIPTION
**Pull request by mlaves**
## Proposed changes

* Implementation of naive `slow_conv_3D` on the cpu.
* Implementation of `explicit_gemm_conv_ND_cpu`.
  * However, using this seems to be considerably slower than the naive implementation. I guess that materializing the strided input view takes very long. The actual gemm is quite fast. Therefore, this is currently unused in conv3d routing.
* Usage of `explicit_gemm_conv_ND_gpu` for GPU implementation of `conv3d`.
* Added tests for `conv3d`.
* Fixed handling of negative padding. The old behavior led to incorrect shapes of the conv gradient w.r.t. the input in some specific kernel size/padding/dilation combinations.
* I also fixed two typos in some error messages I spotted in the code.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Review by jagrit06 — CHANGES_REQUESTED** [review]
Thank you so much for putting this together! Just a few small changes requested and then we can get right to adding it in!

File: mlx/ops.cpp [hunk 1]
```diff
@@ -3077,10 +3101,10 @@ array conv_general(
   // Check for negative padding
   bool has_neg_padding = false;
   for (auto& pd : padding_lo) {
-    has_neg_padding = (pd < 0);
+    has_neg_padding = (pd < 0) || has_neg_padding;
```
[the remark this answers is not in the document]
Comment (mlaves) [reply mlx/ops.cpp:3104]: Done.

File: mlx/primitives.cpp [hunk 2]
```diff
@@ -813,20 +813,62 @@ std::vector<array> Convolution::vjp(
...
+
+      // Use negative padding on the gradient output
+      if (has_neg_padding) {
+        for (auto& p : padding_lo_) {
+          p = std::max(0, p);
+        }
+        for (auto& p : padding_hi_) {
+          p = std::max(0, p);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (mlaves) [reply mlx/primitives.cpp:836]: Yes, this is needed. In some combinations of kernel_size/padding, applying the negative padding before performing the convolution (as done in `conv_general`) leads to wrong gradients w.r.t. the input. E.g., on current master, this will fail without above changes (you need the `|=` fix or it will fail for another reason):
```
import mlx.core as mx
import torch
import numpy as np

x = mx.random.normal((1,8,8,1))
w = mx.random.normal((1,1,1,1))
ct = mx.random.normal((1,5,5,1))

_, grad_mx = mx.vjp(lambda x, w: mx.conv2d(x, w, padding=1, stride=2), [x, w], [ct])
mx_grad_in = grad_mx[0]
pt_grad_in = torch.nn.functional.grad.conv2d_input(
    (1,1,8,8),
    torch.from_numpy(np.array(w)).permute(0,-1,1,2),
    torch.from_numpy(np.array(ct)).permute(0,-1,1,2),
    padding=1,
    stride=2).permute(0, 2, 3, 1)

for i in range(4):
    assert mx_grad_in[i] == pt_grad_in[i], "input grad shape mismatch"
```
One could argue that `kernel_size = 1` and `stride = 2` is unusual, but with my changes, it matches pytorch.

File: mlx/primitives.cpp [hunk 3]
```diff
@@ -813,20 +813,62 @@ std::vector<array> Convolution::vjp(
...
+
+          if (padding_hi[i] < 0) {
+            stops[i + 1] += padding_hi[i];
+            padding_hi[i] = 0;
+          }
+        }
+
+        grad = slice(grad, std::move(starts), std::move(stops), stream());
+      }
+
```
[the remark this answers is not in the document]
Comment (mlaves) [reply mlx/primitives.cpp:871]: See above.

File: mlx/primitives.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jagrit06) [reply mlx/primitives.cpp]: Right, you did add the `|=` fix to `conv_general` though
So what is this block doing now that isn't happening later in the `conv_general` with the `|=` fix ?

[the remark this answers is not in the document]
Comment (mlaves) [reply mlx/primitives.cpp]: The code only passes positive padding to `conv_general` and applies any negative padding afterwards. Without that, the gradient w.r.t. the input will have a wrong shape in some kernel size/stride/padding combinations (as shown in my example).
You can try out to only apply the `|=` fix and then run my example. You’ll see that the input gradient won’t have the same shape as the input, which is clearly incorrect.

[the remark this answers is not in the document]
Comment (jagrit06) [reply mlx/primitives.cpp]: Ooh, I see it now 
The issue is with the way we are doing negative padding when the input is dilated 
I'll write up a general fix for that instead of trying to step around that error

[the remark this answers is not in the document]
Comment (jagrit06) [reply mlx/primitives.cpp]: We can merge this for now though

[the remark this answers is not in the document]
Comment (mlaves) [reply mlx/primitives.cpp]: Ok, fair enough. I fixed the failing test. I think it's ready to merge.

File: mlx/backend/common/conv.cpp [hunk 4]
```diff
@@ -554,6 +868,131 @@ void explicit_gemm_conv_2D_cpu(
   }
 }
 
+void explicit_gemm_conv_ND_cpu(
+    const array& in,
+    const array& wt,
+    array out,
```
Comment (awni) [inline mlx/backend/common/conv.cpp:874]: This function is unused right?

File: mlx/ops.cpp [hunk 5]
```diff
@@ -3035,9 +3059,9 @@ array conv_general(
 
   int spatial_dims = in.ndim() - 2;
 
-  if (spatial_dims < 1 || spatial_dims > 2) {
+  if (spatial_dims < 1 || spatial_dims > 3) {
     throw std::invalid_argument(
-        "[conv] Can only work with inputs that have 1 or 2 spatial dimensions."
+        "[conv] Can only work with inputs that have 1, 2, or 3 spatial dimensions."
```
Comment (awni) [inline mlx/ops.cpp:3064]:
```suggestion
        "[conv] Only works for inputs with 1-3 spatial dimensions."
```

File: python/mlx/nn/layers/convolution.py [hunk 6]
```diff
@@ -132,3 +132,66 @@ def __call__(self, x):
...
+        - ``D`` is the input image depth
+        - ``H`` is the input image height
+        - ``W`` is the input image width
+        - ``C`` is the number of input channels
+    Args:
+        in_channels (int): The number of input channels.
+        out_channels (int): The number of output channels.
+        kernel_size (int or tuple): The size of the convolution filters.
+        stride (int or tuple, optional): The size of the stride when
+            applying the filter. Default: 1.
```
Comment (awni) [inline python/mlx/nn/layers/convolution.py:150]:
```suggestion
            applying the filter. Default: ``1``.
```

File: python/mlx/nn/layers/convolution.py [hunk 7]
```diff
@@ -132,3 +132,66 @@ def __call__(self, x):
...
+        padding (int or tuple, optional): How many positions to 0-pad
+            the input with. Default: 0.
```
Comment (awni) [inline python/mlx/nn/layers/convolution.py:152]:
```suggestion
        padding (int or tuple, optional): How many positions to 0-pad
            the input with. Default: ``0``.
```

File: python/src/ops.cpp [hunk 8]
```diff
@@ -3071,6 +3071,72 @@ void init_ops(nb::module_& m) {
...
+      [](const array& input,
+         const array& weight,
+         const std::variant<int, std::tuple<int, int, int>>& stride,
+         const std::variant<int, std::tuple<int, int, int>>& padding,
+         const std::variant<int, std::tuple<int, int, int>>& dilation,
+         int groups,
+         StreamOrDevice s) {
+        std::tuple<int, int, int> stride_pair{1, 1, 1};
+        std::tuple<int, int, int> padding_pair{0, 0, 0};
+        std::tuple<int, int, int> dilation_pair{1, 1, 1};
```
Comment (awni) [inline python/src/ops.cpp:3084]: The name pair here is a little confusing. Maybe just call them `strides`, `paddings`, `dilations` or something.

**Review by awni — COMMENTED** [review]
This is really nicely done, thank you!

I left a few superficial comments. Would you mind addressing and then we can merge?

**Comment by awni** [comment]
>Implementation of explicit_gemm_conv_ND_cpu.
However, using this seems to be considerably slower than the naive implementation. I guess that materializing the strided input view takes very long. The actual gemm is quite fast. Therefore, this is currently unused in conv3d routing.

Should we delete it? If it's not fast and has no hope of being fast, I don't see much point to leaving it in.

**Comment by awni** [comment]
@mlaves are you planning to update this?

**Comment by mlaves** [comment]
> @mlaves are you planning to update this?

Yes, I will update my PR soon.

**Comment by mlaves** [comment]
> > Implementation of explicit_gemm_conv_ND_cpu.
> > However, using this seems to be considerably slower than the naive implementation. I guess that materializing the strided input view takes very long. The actual gemm is quite fast. Therefore, this is currently unused in conv3d routing.
> 
> Should we delete it? If it's not fast and has no hope of being fast, I don't see much point to leaving it in.

The gemm operation itself is fast, but the reshaping `copy(in_strided_view, in_strided, CopyType::General)` is slow. PyTorch's conv3d CPU implementation is 10x faster [and also uses gemm-conv](https://github.com/pytorch/pytorch/blob/b23b6e7108b8173f9e03805704c8ea0c1e8a2ea5/aten/src/ATen/native/ConvolutionMM3d.cpp#L300). Should I remove this now or is there any chance the reshaping will be faster in the future?

**Comment by awni** [comment]
> Should I remove this now or is there any chance the reshaping will be faster in the future?

It's possible, we haven't looked at / optimized the CPU copies in a while.

**Comment by mlaves** [comment]
> > Should I remove this now or is there any chance the reshaping will be faster in the future?
> 
> It's possible, we haven't looked at / optimized the CPU copies in a while.

In this case, I would keep `explicit_gemm_conv_ND_cpu`, even though it's unused now.

**Comment by awni** [comment]
Sounds good to me!

File: mlx/backend/common/conv.cpp (hunk 4 above)
Comment (mlaves) [reply mlx/backend/common/conv.cpp:874]: See below.

**Comment by mlaves** [comment]
I incorporated all reviewer suggestions.

File: python/src/ops.cpp (hunk 8 above)
Comment (mlaves) [reply python/src/ops.cpp]: Changed it.

**Comment by mlaves** [comment]
I rebased onto main, CI should pass now.
